keps/sig-storage/0031-20181111-extend-datasource-field.md

```diff
@@ -0,0 +1,157 @@
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Drawbacks [optional]](#drawbacks-optional)
+* [Alternatives [optional]](#alternatives-optional)
+
+[Tools for generating]: https://github.com/ekalinin/github-markdown-toc
+
+## Summary
+
+This KEP proposes expanding the allowed uses of the DataSource field for PVC creation in Kubernetes to include PVCs and external Populator CRDs.  Currently we allow Snapshots, which is used by the CSI provisioner to enable creating volumes from snapshots.  In addition to Snaphsots however, the CSI Spec includes the ability to create "clones" (creating a new duplicate volume from an existing volume), in which case the DataSource is an existing Volume on the backend device.  Along with the existing Clone case, the other proposed inclusing would be for a new type of CRD labeleled as a ``Populator``.  Populators are implemented as external CRDs (similar to the Snaphsot CRD) and are comprised of applications that attach a newly created PVC to a POD and populate it with data.
+
+## Motivation
+
+Features like Cloning are common in almost all modern day storage devices, not only is the capability available on most devices, it's also frequently used in various use cases whether it be for duplicating data or to use as a disaster recovery method.  In addition to Clones, one other common process used by storage consumers is to "pre populate" volumes with data, this could be things like restoring a database backup, or seeding a specific data set prior to processing.  In this case it's currently up to the user to implement the population process using their own applications, the concept of an external populator however would greatly simplify this process for the user.
+
+### Goals
+
+* Extend the existing DataSource field for PVCs to allow specifying existing PVCs on the system
+* Extend the existing DataSource field for PVCs to allow specifying external Populator CRDs that are registered on the cluster
+
+
+### Non-Goals
+
+* The actual clone logic is NOT proposed to be implemented in Kubernetes.  This proposal simply proposes the ability to communicate to the CSI Provisioner that we want to perform a clone, it's up to the CSI Plugin/Backend device to handle the implementation.
+* The logic for a Populator would NOT be implemented in Kubernetes.  This would just be a mechanism to indicate that a user would like to use a Populator that's registered on the Cluster as an external CRD.
+
+## Proposal
+
+Modify the checks on the dataSource field to allow:
+1. Existing PVCs in the users namespace (clones).
+2. External Populators in the form of registered CRDs
+
+In both cases the proposal is just to allow communication of intent, there's no implementation proposed in Kubernetes.  The Clone case is fairly straight forward given that the CSI spec already includes volumes as a data source and has a cloning capability that plugins are required to report:
```

Comment:
External-provisioner should be mentioned here.  In case 1, it will call the CSI plugin to create a volume with the PVC data source.  In case 2, I think the external-provisioner will call the CSI plugin to create an empty volume, but then data will be populated later by the populator.

After reading the remaining doc, I see that you did talk about the role of the CSI provisioner.  Maybe add a brief description here too.  When I read "there's no implementation proposed in Kubernetes", I thought you are not going to mention the external provisioner which is under kubernetes-csi:).